p2p/wdrr_queue.go

```diff
@@ -0,0 +1,301 @@
+package p2p
+
+import (
+	"github.com/gogo/protobuf/proto"
+	"github.com/tendermint/tendermint/libs/log"
+	tmsync "github.com/tendermint/tendermint/libs/sync"
+)
+
+type drrFlow struct {
+	mtx    tmsync.RWMutex
```

Comment:
`RWMutex` is much slower than `Mutex` (2-3x if I remember correctly), so is only worth it in aggregate with many concurrent readers, which I don't believe we have here.

If possible, it would also be awesome if we could have a single scheduling goroutine, such that we wouldn't need a mutex at all -- although this needs to be measured before starting to optimize (performance may be dominated by other aspects).